tests/kafkatest/services/kafka/kafka.py

```diff
@@ -89,11 +91,38 @@ def start(self):
                 topic_cfg["topic"] = topic
                 self.create_topic(topic_cfg)
 
+    def set_protocol_and_port(self, node):
```

Comment:
It's a little confusing to me right now what `use_separate_ports_per_protocol` and `force_both_protocols_open` should mean if other security protocols are in play (`SASL_PLAINTEXT` or `SASL_SSL`), and overall having a lot of conditionals is a bit complicated.

The main idea seems to be to dedicate one port number to one security protocol, and then allow for a given port being open or closed.

I was just brainstorming how this section might be simplified or removed. See if this idea makes sense to you:

Add the concept of a port, which can be open or closed (True or False), so something like:
```
Port = namedtuple('Port', ['name', 'number', 'open']) 
# e.g. p = Port('PLAINTEXT', 9092, False)
``` 

KafkaService then has a ports field which is a mapping from security protocol to `Port` object, so something like:
```
self.ports = {
  'PLAINTEXT': Port('PLAINTEXT', 9092, False),
  'SSL': Port('SSL', 9093, False),
  'SASL_PLAINTEXT': Port('SASL_PLAINTEXT', 9094, False),
  'SASL_SSL': Port('SASL_SSL', 9095, False)
}
```

The KafkaService constructor could do something like this:
```
self.ports[security_protocol].open = True
self.ports[interbroker_security_protocol].open = True
```

Similary, in the course of the rolling upgrade, you could toggle the corresponding port on or off. 

As I understand things, this could get rid of the complicated conditionals, and hopefully make the intent of the code a bit clearer.